### PR TITLE
[megatron] fused linear cross-entropy for 23x memory savings at 262k context

### PR DESCRIPTION
**Pull request by casper-hansen**
I wanted to run Qwen/Qwen3.6-35B-A3B on 8xB300 using SkyRL and found that it cannot handle full context without OOMing on the cross-entropy step in the trainer. To solve this, I implemented a fused cross-entropy that saves a lot of memory for the Megatron backend while handling the parallelism.

## Why not the existing Megatron/Apex/TE fused cross-entropy?

`fused_vocab_parallel_cross_entropy` optimizes the CE *compute*, not *memory*: it materializes the `[S, vocab//TP]` logits and its staged fused kernels allocate extra buffers, so it uses ~1.7x **more** memory than eager CE and OOMs earlier at long context.

### Per-rank vocab = 62,080 (the real TP=4 shard)

| seq_len | baseline (eager CE) | nvidia (fused CE) | fused linear CE (ours) | fused linear mem vs baseline |
|---:|---:|---:|---:|---:|
| 8,192 | 3.1 GB / 11 ms | 5.0 GB / 8 ms | 2.1 GB / 25 ms | 1.5× |
| 32,768 | 11.7 GB / 43 ms | 19.3 GB / 29 ms | 2.3 GB / 98 ms | 5.1× |
| 65,536 | 23.2 GB / 86 ms | 38.4 GB / 58 ms | 2.5 GB / 196 ms | 9.1× |
| 131,072 | 46.2 GB / 175 ms | 76.5 GB / 120 ms | 3.0 GB / 392 ms | 15.2× |
| 262,144 | 92.2 GB / 357 ms | 152.8 GB / 245 ms | 4.0 GB / 784 ms | 22.8× |

### Full vocab = 248,320

| seq_len | baseline (eager CE) | nvidia (fused CE) | fused linear CE (ours) |
|---:|---:|---:|---:|
| 8,192 | 12.4 GB / 42 ms | 19.9 GB / 34 ms | 8.1 GB / 92 ms |
| 32,768 | 46.6 GB / 177 ms | 76.9 GB / 151 ms | 8.3 GB / 367 ms |
| 65,536 | 92.2 GB / 356 ms | 152.8 GB / 316 ms | 8.6 GB / 734 ms |
| 131,072 | 183.3 GB / 723 ms | **OOM** | 9.1 GB / 1,467 ms |
| 262,144 | **OOM** | **OOM** | 10.1 GB / 2,932 ms |

## Convergence

I did a small SFT test to overfit on GSM8K with Qwen/Qwen2.5-1.5B-Instruct. Convergence is the same, just with slight variations as the computation has some natural variance when it's chunked.

image

### CODE REVIEW

File: skyrl/backends/skyrl_train/distributed/megatron/model_utils.py [hunk 1]
```diff
@@ -265,6 +265,152 @@ def backward(
...
+            log_probs[target_mask[:, chunk_start:chunk_end]] = 0.0
+
+            torch.distributed.all_reduce(
+                log_probs,
+                op=torch.distributed.ReduceOp.SUM,
+                group=tp_group,
+            )
+            all_log_probs.append(log_probs)
+
+        log_probs = torch.cat(all_log_probs, dim=1)
```
[the remark this answers is not in the document]
Comment (casper-hansen) [reply skyrl/backends/skyrl_train/distributed/megatron/model_utils.py]: Good call, done. I moved the `all_reduce` out of the chunk loop so it runs once on the full `[B, S]` logprob tensor after the cat instead of once per chunk. Each rank only holds the logprob for its own vocab shard (0 elsewhere) and the cross-TP combine is just a SUM, so concatenating first and reducing once gives the identical result. That takes it from `num_chunks` collectives down to 1 (256 -> 1 at 262k with chunk=1024).

File: skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py [hunk 2]
```diff
@@ -89,6 +124,10 @@ def __init__(
         self.actor_optimizer = actor_optimizer
         self.policy_loss_fn = policy_loss_fn
         self.remove_microbatch_padding = self.cfg.remove_microbatch_padding
+        # Fuse the LM-head projection into the chunked log-prob/entropy via the
+        # GPTModel output_processor hook (avoids materializing the full
+        # [B, S, vocab//TP] logits + its fp32 grad). See model_utils.
+        self._fused_lm_head = bool(getattr(self.cfg, "fused_lm_head_logprob", False))
```
[the remark this answers is not in the document]
Comment (casper-hansen) [reply skyrl/backends/skyrl_train/workers/megatron/megatron_model_wrapper.py:128]: Yeah, good catch. The forward-only path was still materializing the full logits. I wired the fused head into it too: `forward()` now passes the same `output_processor`, and its collection_func uses `from_parallel_hidden_to_logprobs(inference_only=True)` when the fused head is on, so the reference/old-logprob passes go through the chunked path and don't OOM at long context.

File: skyrl/backends/skyrl_train/distributed/megatron/model_utils.py [hunk 3]
```diff
@@ -265,6 +265,152 @@ def backward(
...
+            grad_output_selected = chunk_grad_output.masked_select(valid_mask)
+            grad_logits.view(-1).scatter_add_(0, flat_chosen, grad_output_selected)
+
+            grad_logits = grad_logits.to(dtype=weight.dtype)  # [B, cs, V//TP]
+
+            # Project chunk logit-grad back to hidden / weight grads.
+            grad_hidden[:, chunk_start:chunk_end, :] = torch.matmul(grad_logits, weight)
+            grad_logits_2d = grad_logits.reshape(-1, partition_vocab_size)
+            h_2d = h_chunk.reshape(-1, hidden_size)
+            grad_weight.add_(torch.matmul(grad_logits_2d.t().to(torch.float32), h_2d.to(torch.float32)))
```
[the remark this answers is not in the document]
Comment (casper-hansen) [reply skyrl/backends/skyrl_train/distributed/megatron/model_utils.py]: Done. I run the matmul in `grad_logits.dtype` (the compute/weight dtype) now and only upcast the result to fp32 for the accumulation into `grad_weight`, which matches what `ColumnParallelLinear` does in its own backward. Nice side effect: the long-context backward got about 2x faster (the 262k per-rank shard went from ~1700ms to ~780ms) at the same peak memory.

**Comment by dyurk-lila** [comment]
How does this compare with https://github.com/NovaSky-AI/SkyRL/pull/1765?

**Comment by casper-hansen** [comment]
> How does this compare with #1765?

I don't know, you didn't include any benchmarks in your PR. Please feel free to compare! In terms of simplicity, this one gives you a lot for 1k lines of code where as yours seem to import from verl with 3.5k lines of code.

**Comment by casper-hansen** [comment]
@SumanthRH @erictang000 Could you help review? Much appreciated!

**Comment by casper-hansen** [comment]
GSM8K RL GRPO run (another convergence test)

image

**Comment by erictang000** [comment]
hey @dyurk-lila @casper-hansen 

thanks both for the PRs - so the current overlap is that both of the two PRs #1841 and #1765 are adding a fused + chunked lm head.

#1841 (this pr) adds:
- a fused torch only implementation of the chunked LM head, with an additional chunked entropy calculation/helper that currently enforces no grad (ran some of my own claude code validation here: https://github.com/erictang000/SkyRL/blob/pr-1841/FUSED_LM_HEAD_PROFILE_REPORT.md)

#1765 adds:
- a fused torch only implementation 
- a fused triton implementation from verl that shows speedups over torch only
- both of which compute entropy along with logprobs at the same time, with grads, saving an additional forward pass for entropy. 

I'm inclined to merge #1841 first, since it's less LOC and less invasive for now, then we can gradually port over/merge components from #1765, in order:
1. entropy backward for fused torch implementation
2. triton backend kernel + efficient entropy kernel + backend selection ("torch" vs "triton")

**Review by erictang000 — APPROVED** [review]
very clean PR, lgtm! thanks again @casper-hansen
